### PR TITLE
[5486] - withdrawal reasons migration

### PR DESCRIPTION
**Pull request by d-a-v-e**
### Context

This covers the work needed to migrate over to the new withdraw reasons.

- new models/tables WithdrawlReasons and TraineeWithdrawReasons
- data migration for existing withdraw reasons
- mapping of legacy withdraw reasons to new reasons
- storing legacy reasons with no direct mappings
- dropping old withdraw_reason enum attribute
- fixing db/seed to work with new schema
- DO NOT MERGE TO MAIN - interim PR until greater bulk of withdraw flow is built

### CODE REVIEW

File: db/data/20230515083653_trainee_withdrawal_reasons.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class TraineeWithdrawalReasons < ActiveRecord::Migration[7.0]
+  def up
+    # Create TraineeWithdrawalReason entries for each withdrawal reason
+    Trainee.withdraw_reason.each_key do |withdraw_reason|
```
Comment (jhenderson) [inline db/data/20230515083653_trainee_withdrawal_reasons.rb:6]:
```suggestion
    Trainee.withdraw_reasons.each_key do |withdraw_reason|
```

File: db/migrate/20230512103034_create_withdrawal_reasons.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+class CreateWithdrawalReasons < ActiveRecord::Migration[7.0]
+  def change
+    create_table :withdrawal_reasons do |t|
+      t.timestamps
```
Comment (jhenderson) [inline db/migrate/20230512103034_create_withdrawal_reasons.rb:6]: Suggest putting `t.timestamps` at the end

File: db/seeds.rb [hunk 3]
```diff
@@ -24,6 +24,14 @@
   unique_by: :name,
 )
 
+# load Withdraw reasons
```
Comment (jhenderson) [inline db/seeds.rb:27]: No need to set timestamps. It's automatically done as you know.

`WithdrawalReason.upsert_all(WithdrawalReasons::SEED)`

File: db/data/20230515083653_trainee_withdrawal_reasons.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+      # Find the corresponding withdrawal_reason_id
+      withdrawal_reason_id = WithdrawalReason.find_by(name: mapped_reason).id
+
+      # Find all trainees with this withdrawal reason
+      trainees = Trainee.where(withdraw_reason:)
+      next unless trainees.any?
+
+      # Create TraineeWithdrawalReason entries
+      TraineeWithdrawReason.create(
```
Comment (jhenderson) [inline db/data/20230515083653_trainee_withdrawal_reasons.rb:18]:
```suggestion
      TraineeWithdrawalReason.create(
```

File: db/data/20230515083653_trainee_withdrawal_reasons.rb [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+class TraineeWithdrawalReasons < ActiveRecord::Migration[7.0]
+  def up
```
Comment (jhenderson) [inline db/data/20230515083653_trainee_withdrawal_reasons.rb:4]: Run `WithdrawalReason.upsert_all(WithdrawalReasons::SEED)` here before data migration

File: db/seeds.rb (hunk 3 above)
Comment (d-a-v-e) [reply db/seeds.rb:27]: oh cool!

**Pull request merged**